aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -329,4 +355,234 @@ CUDAGraph::~CUDAGraph() {
+      "get_currently_capturing_graph() can be used only between capture_begin() and capture_end()");
+  return _currently_capturing_graph;
+}
+
+void CUDAGraph::begin_capture_to_if_node(
+    const at::Tensor& scalar_cuda_pred_tensor) {
+#if !defined(USE_ROCM) && (defined(CUDA_VERSION) && CUDA_VERSION >= 12040)
+  TORCH_CHECK(
+      !has_graph_exec_,
+      "begin_capture_to_if_node() must be called before capture_begin()");
+
+  cudaStreamCaptureStatus status;
+  cudaGraph_t currently_capturing_graph;
+  AT_CUDA_CHECK(cudaStreamGetCaptureInfo(
+      getCurrentCUDAStream(), &status, nullptr, &currently_capturing_graph));
+  TORCH_CHECK(
+      status == cudaStreamCaptureStatusActive,
+      "capture_begin() must be called before begin_capture_to_if_node()");
+  cudaGraphConditionalHandle handle;
+  AT_CUDA_CHECK(cudaGraphConditionalHandleCreate(
+      &handle, currently_capturing_graph, 0, 0));
+
+  set_conditional_handle(handle, scalar_cuda_pred_tensor);
+
+  const cudaGraphNode_t* dependencies;
+  size_t num_dependencies;
+  AT_CUDA_CHECK(cudaStreamGetCaptureInfo(
+      getCurrentCUDAStream(),
+      &status,
+      nullptr,
+      &currently_capturing_graph,
+      &dependencies,
+      &num_dependencies));
+  TORCH_CHECK(status == cudaStreamCaptureStatusActive);
+
+  cudaGraphNodeParams params{};
+  params.type = cudaGraphNodeTypeConditional;
+  params.conditional.handle = handle;
+  params.conditional.type = cudaGraphCondTypeIf;
+  params.conditional.size = 1;
+
+  cudaGraphNode_t cond_node;
+  AT_CUDA_CHECK(cudaGraphAddNode(
+      &cond_node,
+      currently_capturing_graph,
+      dependencies,
+      num_dependencies,
+      &params));
+
+  cudaGraph_t if_node_child_graph = params.conditional.phGraph_out[0];
+
+  AT_CUDA_CHECK(cudaStreamUpdateCaptureDependencies(
+      getCurrentCUDAStream(), &cond_node, 1, cudaStreamSetCaptureDependencies));
+
+  // From:
+  // https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g793d7d4e474388ddfda531603dc34aa3
+  // "Capture must be ended on the same stream in which it was initiated, and it
+  // may only be initiated if the stream is not already in capture mode."
+
+  // Since pytorch uses a pool of 32 pre-allocated cuda streams,
+  // should a user nest 32 conditional nodes, there would be an error
+  // for the 32nd node, since that node's stream would already be in
+  // capture mode. The easiest solution is to handle stream creation
+  // and deletion ourselves.
+  cudaStream_t child_stream;
+  AT_CUDA_CHECK(cudaStreamCreate(&child_stream));
+  conditional_graph_capture_streams_ids_.push(-1);
+  c10::cuda::CUDACachingAllocator::endAllocateToPool(capture_dev_, mempool_id_);
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
+      capture_dev_, mempool_id_, create_child_allocate_filter());
+  AT_CUDA_CHECK(cudaStreamBeginCaptureToGraph(
+      child_stream, if_node_child_graph, nullptr, nullptr, 0, capture_mode_));
```

Comment:
This code is actually used to capture either true_block or false_block.

CUDA currently does not currently support an if-else conditional nodes (this is coming out in a later version of cuda, so it is a few months out), so as I mention elsewhere, I use two if conditional nodes, one for true_block and one for false_block, in order to implement torch.cond.

When you ask "are we capturing cond", I'm not sure if you refer to capturing the entirety of torch.cond or evaluation of the boolean predicate. But basically, the entirety of torch.cond will run in the stream that was being used when begin_capture_to_if_node() was called. The actual execution does not need to use either of the ancillary streams used to capture true_block or the false_block, since only one is going to execute. Therefore, the original cuda stream will be used, and thus there is no need for a "channel switch", which can cost overhead. It's just a quirk of implementation that a new stream is required to create the true_block and the fasle_block.